0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+  /// the nodes in the new committee.
+  ///
+  /// Full share:
+  ///      s'(y) = B'(node_id,y)
+  fn full_share_distribution(&self, churp_id: u8, round: u64);
+
+  /// Return switch data point (integer) for the calling node.
+  /// 
+  /// The point is evaluation of the shared secret (polynomial)
+  /// at the given y value (caller id).
+  /// 
+  /// Switch point:
+  ///     P = g(caller_id) = B(node_id,caller_id)
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn switch_point(&self, churp_id: u8, round: u64) -> Integer;
+
+  /// Return the key share for the given round.
+  ///
+  /// To construct a derived key, the caller is required to collect 
+  /// at least the threshold number of key shares and combine them
+  /// locally to construct the derived key.
+  ///
+  /// Key share:
+  ///     K = H(runtime_id, key_pair_id)^{r g(0)}
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the key share needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn key_share(&self, churp_id: u8, round: u64, runtime_id: Namespace, 
+    key_pair_id: KeyPairId) -> Integer;
+}
+```
+
+Methods:
+
+- `churp/init`
+
+- `churp/verification_matrix`
+
+- `churp/bivariate_share`
+
+- `churp/dealing`
+
+- `churp/share_reduction`
+
+- `churp/proactive_randomization`
+
+- `churp/full_share_distribution`
+
+- `churp/switch_point`
+
+- `churp/key_share`
+
+### Key manager client
+
+```rust
+/// Key manager client for CHURP.
+pub trait Client {
+  /// Return the key for the given round.
+  ///
+  /// The key is combined from key shares obtained from the key manager nodes
+  /// in the current CHURP committee.
+  fn key(&self, churp_id: u8, round: u64, runtime_id: Namespace, key_pair_id: KeyPairId) -> Vec<u8>;
+}
+```
+
+### Positive (CHURP)
+
+- High security, as the master secret is shared among key manager nodes.
+- Supports proactive randomization (share refresh).
+- Dynamic committees.
+
+### Negative (CHURP)
+
+- Slower than KDC.
+
+## Key Derivation Center
+
+Key Derivation Center (KDC) is a secret sharing scheme based on
+the verifiable secret sharing scheme (VSS) where every node possesses
+only a share of the master secret. To derive a key from the master secret,
+one needs to obtain at least t+1 threshold number of key shares from distinct
+nodes and reassemble them locally.
+
+```rust
+/// Key manager application that implements key derivation center (KDC).
+pub trait KDC {
+  /// TODO: Define methods if we decide to implement KDC also.
+}
+```
+
+### Positive (KDC)
+
+- High security, as the master secret is shared among key manager nodes.
+- Supports proactive randomization (share refresh).
+- Faster than CHURP.
+
+### Negative (KDC)
+
+- The number of key manager nodes that share a master secret is fixed and
+  cannot be changed once shares are generated.
```

Comment:
Does that mean that if a key manager node is destroyed the remaining committee can never reshare to introduce a replacement node?

Reply:
Yes. If you use (t, n)-scheme, then once n-t+1 nodes are destroyed, the secret is lost.

Reply:
Ok, I would explicitly mention this here.